### PR TITLE
Pt 165857097 add gas and value to calls

### PR DESCRIPTION
**Pull request by gorillainduction**
https://www.pivotaltracker.com/story/show/165857097

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -62,7 +62,7 @@
                    {ref, "9ff46af"}}},
 
         {aebytecode, {git, "https://github.com/aeternity/aebytecode.git",
-                     {ref,"2555868"}}},
+                     {ref,"2f4e1888"}}},
```
Comment (hanssv) [inline rebar.config]: Why 8 characters?

File: apps/aefate/src/aefa_fate.erl [hunk 2]
```diff
@@ -374,18 +397,46 @@ push(V, ES) ->
...
+            pop_call_stack(ES2);
+        [{Contract, Function, BB, Mem, Value}| Rest] ->
+            ES0 = aefa_engine_state:set_call_value(Value, ES),
+            ES1 = aefa_engine_state:set_call_stack(Rest, ES0),
             ES2 = aefa_engine_state:set_memory(Mem, ES1),
             ES3 = set_function(Contract, Function, ES2),
             {jump, BB, ES3}
     end.
 
+abort_out_of_gas(ES) ->
```
Comment (UlfNorell) [inline apps/aefate/src/aefa_fate.erl]: Doesn't actually abort, so call it something else?

**Comment by gorillainduction** [comment]
Rebase to resolve conflicts

File: apps/aefate/src/aefa_fate_op.erl [hunk 3]
```diff
@@ -138,34 +141,75 @@ returnr(Arg0, EngineState) ->
...
+    {Fun, ES1} = get_op_arg(Arg0, EngineState),
     Signature = aefa_fate:get_function_signature(Fun, ES1),
-    {ok, ES2} = aefa_fate:check_signature_and_bind_args(Signature, ES1),
+    ok = aefa_fate:check_signature(Signature, ES1),
+    ES2 = aefa_fate:bind_args_from_signature(Signature, ES1),
     {jump, 0, aefa_fate:set_local_function(Fun, ES2)}.
 
-call_r(Arg0, Arg1, EngineState) ->
+call_r(Arg0, Arg1, Arg2, EngineState) ->
+    Sender = aefa_engine_state:current_contract(EngineState),
```
Comment (hanssv) [inline apps/aefate/src/aefa_fate_op.erl]: Could be done in `call_r_common`?

File: apps/aefate/src/aefa_fate_op.erl [hunk 4]
```diff
@@ -138,34 +141,75 @@ returnr(Arg0, EngineState) ->
...
+    ES5 = transfer_value(Sender, Address, Value, ES4),
+    {jump, 0, ES5}.
 
-call_t(Arg0, EngineState) ->
-    {Fun, ES0} = get_op_arg(Arg0, EngineState),
-    Signature = aefa_fate:get_function_signature(Fun, ES0),
-    {ok, ES2} = aefa_fate:check_signature_and_bind_args(Signature, ES0),
-    {jump, 0, aefa_fate:set_local_function(Fun, ES2)}.
+call_gr(Arg0, Arg1, Arg2, Arg3, EngineState) ->
+    Sender = aefa_engine_state:current_contract(EngineState),
```
Comment (hanssv) [inline apps/aefate/src/aefa_fate_op.erl]: Could be done in `call_gr_common`?

File: rebar.config (hunk 1 above)
Comment (gorillainduction) [reply rebar.config]: Fumbly fingers? No reason whatsoever.

File: apps/aefate/src/aefa_fate_op.erl (hunk 3 above)
Comment (gorillainduction) [reply apps/aefate/src/aefa_fate_op.erl]: :thinking: Yes. It just needs to happen before the binding of new arguments.

File: apps/aefate/test/aefa_sophia_test.erl [hunk 5]
```diff
@@ -62,6 +74,7 @@ make_call(Contract, Function, Arguments) ->
     Calldata = {tuple, {Function, {tuple, EncArgs}}},
     #{ contract => pad_contract_name(Contract),
        gas      => 1000000,
+       value    => 10000,
        call     => aeb_fate_encoding:serialize(Calldata) }.
```
Comment (UlfNorell) [inline apps/aefate/test/aefa_sophia_test.erl]: Might consider adding a `Call.value` to the remote test. Perhaps
```
  function remote(x : int) = x * (2 + Call.value)
```

File: apps/aefate/src/aefa_fate.erl (hunk 2 above)
Comment (gorillainduction) [reply apps/aefate/src/aefa_fate.erl]: Sure. It used to do abort in the end ;-)

File: apps/aefate/test/aefa_sophia_test.erl (hunk 5 above)
Comment (gorillainduction) [reply apps/aefate/test/aefa_sophia_test.erl]: I would if the compiler could handle it ;-)
I see another PR in the future.

Comment (hanssv) [reply apps/aefate/test/aefa_sophia_test.erl]: 🔮

**Pull request merged**